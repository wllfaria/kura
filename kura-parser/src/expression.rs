use kura_lexer::token::{Kind, Location, Operator, Primitive, Token, Value};
use kura_lexer::{Lexer, TransposeRef};

use crate::Expression;

mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 3;
    pub const MUL: u8 = 4;
    pub const ASSOC: u8 = 5;
    pub const APPLY: u8 = 6;
}

fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::Star | Operator::Slash => precedences::MUL,
        Operator::And => precedences::ASSOC,
        Operator::LeftParen | Operator::EqualEqual | Operator::NotEqual => precedences::APPLY,
        _ => precedences::BASE,
    }
}

pub fn parse_expression<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    match lexer.peek().transpose().map_err(ToString::to_string)? {
        Some(token) => match token.kind {
            Kind::Var | Kind::Const => parse_variable(lexer),
            _ => parse_with_precedence(lexer, precedences::BASE, base_expr),
        },
        None => unreachable!(),
    }
}

pub fn parse_identifier<'parser>(lexer: &mut Lexer<'parser>) -> Result<(Expression<'parser>, &'parser str), String> {
    let name_and_loc = lexer
        .next()
        .transpose()
        .map_err(|e| e.to_string())?
        .map(|token| match token.kind {
            Kind::Value(Value::Ident(name)) => (name, token.location),
            _ => ("", token.location),
        });

    match name_and_loc {
        Some(("", location)) => Err(location.to_string()),
        Some((name, location)) => Ok((Expression::Ident { name, location }, name)),
        None => Err("TODO".into()),
    }
}

fn parse_expr_block<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    let mut expressions = vec![];

    let block_start = lexer.expect(Kind::Op(Operator::LeftBrace)).map_err(|e| e.to_string())?;

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightBrace)) => {
                break;
            }
            None => break,
            _ => (),
        }

        let expr = parse_expression(lexer, base_expr)?;
        expressions.push(expr);
    }

    let block_end = lexer
        .expect(Kind::Op(Operator::RightBrace))
        .map_err(|e| e.to_string())?;

    let location = block_start.location.start_byte..block_end.location.end_byte;
    Ok(Expression::Block {
        expressions,
        location: location.into(),
    })
}

fn parse_variable<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer
        .expect_one_of(&[Kind::Var, Kind::Const])
        .map_err(|e| e.to_string())?;
    let mutable = matches!(keyword.kind, Kind::Var);

    let (_, name) = parse_identifier(lexer)?;

    let typ = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::Colon)) => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            Some(parse_identifier(lexer)?)
        }
        _ => None,
    };

    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match token.kind {
            Kind::Op(Operator::LeftBrace) => parse_expr_block(lexer, false)?,
            _ => parse_expression(lexer, false)?,
        },
        _ => unreachable!(),
    };

    lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = Location::new(keyword.location.start_byte, value.location().end_byte);
    Ok(Expression::Var {
        mutable,
        typ: typ.map(|(typ, _)| Box::new(typ)),
        name,
        value: Box::new(value),
        location,
    })
}

fn parse_if_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::If).map_err(|e| e.to_string())?;

    let condition = parse_expression(lexer, false)?;

    let body = parse_expr_block(lexer, true)?;

    let mut falsy_branches = vec![];

    while let Some(keyword) = lexer.peek().transpose().map_err(|e| e.to_string())? {
        if keyword.kind == Kind::Else {
            // consume `else`
            lexer.next().transpose().map_err(|e| e.to_string())?;

            if lexer
                .peek()
                .transpose()
                .map_err(|e| e.to_string())?
                .map(|t| matches!(t.kind, Kind::If))
                .unwrap_or(false)
            {
                let else_if = parse_if_expression(lexer)?;
                falsy_branches.push(else_if);
            } else {
                let else_block = parse_expr_block(lexer, true)?;
                falsy_branches.push(else_block);
                break;
            }
        } else {
            break;
        }
    }

    let end = match falsy_branches.last() {
        Some(branch) => branch.location().end_byte,
        None => body.location().end_byte,
    };

    let location = keyword.location.start_byte..end;
    Ok(Expression::If {
        condition: Box::new(condition),
        truthy: Box::new(body),
        falsy: falsy_branches,
        location: location.into(),
    })
}

fn parse_value<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Value(value) = kind else {
        unreachable!();
    };

    match value {
        Value::Primitive(_) => parse_primitive(lexer),
        Value::Ident(_) => Ok(parse_identifier(lexer)?.0),
        t => todo!("{t}"),
    }
}

fn parse_operation<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Op(op) = kind else {
        unreachable!();
    };

    match op {
        Operator::LeftParen => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            let left = parse_with_precedence(lexer, precedences::BASE, false)?;
            lexer
                .expect(Kind::Op(Operator::RightParen))
                .map_err(|e| e.to_string())?;
            Ok(left)
        }
        Operator::LeftBrace => parse_expr_block(lexer, true),
        t => todo!("{t:?}"),
    }
}

fn parse_fun_call<'parser>(
    lexer: &mut Lexer<'parser>,
    ident: Expression<'parser>,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::LeftParen)).map_err(|e| e.to_string())?;

    let mut arguments = vec![];

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightParen)) => {
                break;
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Comma)) => {
                lexer.next().transpose().map_err(|e| e.to_string())?;
                continue;
            }
            None => break,
            _ => (),
        }

        let arg = parse_expression(lexer, false)?;

        arguments.push(arg);
    }

    let close_paren = lexer
        .expect(Kind::Op(Operator::RightParen))
        .map_err(|e| e.to_string())?;

    if base_expr {
        lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;
    }

    let location = ident.location().start_byte..close_paren.location.end_byte;
    let expr = Expression::FunCall {
        ident: Box::new(ident),
        location: location.into(),
        arguments,
    };
    Ok(expr)
}

fn parse_assign<'parser>(lexer: &mut Lexer<'parser>, left: Expression<'parser>) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::LeftBrace)) => parse_expr_block(lexer, false)?,
        Some(_) => parse_expression(lexer, false)?,
        _ => unreachable!(),
    };

    let closing = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = left.location().start_byte..closing.location.end_byte;

    Ok(Expression::Assign {
        ident: Box::new(left),
        location: location.into(),
        value: Box::new(value),
    })
}

fn parse_with_precedence<'parser>(
    lexer: &mut Lexer<'parser>,
    min_precedence: u8,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    let mut left = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match &token.kind {
            Kind::Value(_) => parse_value(lexer)?,
            Kind::Op(_) => parse_operation(lexer)?,
            Kind::Return => parse_return_expression(lexer)?,
            Kind::If => parse_if_expression(lexer)?,
            t => todo!("{t:?}"),
        },
        None => todo!(),
    };

    if let Expression::Ident { .. } = left {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::LeftParen)) => {
                return parse_fun_call(lexer, left, base_expr);
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Equal)) => return parse_assign(lexer, left),
            _ => (),
        }
    }

    loop {
        let Some(next) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
            return Ok(left);
        };

        let Kind::Op(operator) = next.kind else {
            return Ok(left);
        };

        if !next.kind.is_binary_op() {
            return Ok(left);
        }

        let precedence = get_precedence(operator);

        if precedence <= min_precedence {
            break;
        }

        let Some(_) = lexer.next().transpose().map_err(|e| e.to_string())? else {
            unreachable!();
        };

        let right = parse_with_precedence(lexer, precedence, false)?;

        let location = Location::new(left.location().start_byte, right.location().end_byte);
        left = Expression::BinaryOp {
            lhs: Box::new(left),
            operator,
            rhs: Box::new(right),
            location,
        };
    }

    Ok(left)
}

fn parse_return_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::Return).map_err(|e| e.to_string())?;

    let value = parse_expression(lexer, false)?;

    let ending_semi = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = keyword.location.start_byte..ending_semi.location.end_byte;
    Ok(Expression::Return {
        value: Box::new(value),
        location: location.into(),
    })
}

fn parse_primitive<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let (primitive, location) = match lexer.next().transpose().map_err(|e| e.to_string())? {
        Some(Token {
            kind: Kind::Value(Value::Primitive(primitive)),
            location,
            ..
        }) => (primitive, location),
        _ => unreachable!(),
    };

    match primitive {
        Primitive::Int { value, size } => Ok(Expression::IntLiteral { value, size, location }),
        Primitive::UInt { value, size } => Ok(Expression::UintLiteral { value, size, location }),
        Primitive::Float { value, size } => Ok(Expression::FloatLiteral { value, size, location }),
        Primitive::Bool(value) => Ok(Expression::Bool { value, location }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Parser;

    fn make_sut(source: &str) -> Parser<'_> {
        let lexer = Lexer::new(source);
        Parser::new(source, lexer)
    }

    #[test]
    fn simple_math() {
        let simple_math_expr = "1 + 2 * 3 - 4";

        let mut parser = make_sut(simple_math_expr);

        let math_expr_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(math_expr_ast);
    }

    #[test]
    fn mutable_variable() {
        let variables = "var hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn immutable_constant() {
        let variables = "const hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn if_statement() {
        let source = r#"
            if something == another_thing && 1 != 2 {
                const this_is_a_var = 10 + 3;
            } else if something == 10 {
                const this_is_another = 10 + 10;
            } else {
                const omg = 1 + 1;
            }
        "#;
        let mut parser = make_sut(source);

        let if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(if_ast);
    }

    #[test]
    fn if_as_variable_value() {
        let source = r#"
            const my_var = if something == another_thing {
                10 + 10
            } else {
                20 + 20
            };
        "#;

        let mut parser = make_sut(source);
        let let_if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(let_if_ast);
    }
}
