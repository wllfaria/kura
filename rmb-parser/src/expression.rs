use miette::{Error, LabeledSpan};

use crate::Expression;
use rmb_lexer::{
    token::{Kind, Location, Operator, Primitive, Token, Value},
    Lexer, TransposeRef,
};

mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 3;
    pub const MUL: u8 = 4;
    pub const APPLY: u8 = 5;
}

fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::Star | Operator::Slash => precedences::MUL,
        Operator::LeftParen => precedences::APPLY,
        _ => precedences::BASE,
    }
}

pub fn parse_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, Error> {
    match lexer.peek().transpose()? {
        Some(token) => match token.kind {
            Kind::Var | Kind::Const => parse_variable(lexer),
            _ => parse_with_precedence(lexer, precedences::BASE),
        },
        None => unreachable!(),
    }
}

pub fn parse_identifier<'parser>(
    lexer: &mut Lexer<'parser>,
) -> Result<(Expression<'parser>, &'parser str), Error> {
    let name_and_loc = lexer.next().transpose()?.map(|token| match token.kind {
        Kind::Value(Value::Ident(name)) => (name, token.location),
        _ => ("", token.location),
    });

    match name_and_loc {
        Some(("", location)) => Err(miette::miette! {
            labels = vec![
                LabeledSpan::at(location.start_byte..location.end_byte, "this identifier"),
            ],
            "",
        }
        .with_source_code(lexer.complete_source.to_string())),
        Some((name, location)) => Ok((Expression::Ident { name, location }, name)),
        None => {
            let location = lexer.complete_source.len() - 1..lexer.complete_source.len();
            Err(miette::miette! {
                labels = vec![
                    LabeledSpan::at(location, "at this location"),
                ],
                "unexpected end of file",
            }
            .with_source_code(lexer.complete_source.to_string()))
        }
    }
}

fn parse_expr_block<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, Error> {
    let mut expressions = vec![];

    let block_start = lexer.expect(Kind::Op(Operator::LeftBrace))?;

    loop {
        match lexer.peek().transpose()? {
            Some(token) => {
                if let Kind::Op(Operator::RightBrace) = token.kind {
                    break;
                }
            }
            None => break,
        }

        let expr = parse_expression(lexer)?;
        expressions.push(expr);
    }

    let block_end = lexer.expect(Kind::Op(Operator::RightBrace))?;

    let location = block_start.location.start_byte..block_end.location.end_byte;
    Ok(Expression::Block {
        expressions,
        location: location.into(),
    })
}

fn parse_variable<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, Error> {
    let keyword = lexer.expect_one_of(&[Kind::Var, Kind::Const])?;
    let mutable = matches!(keyword.kind, Kind::Var);

    let (_, name) = parse_identifier(lexer)?;

    lexer.expect(Kind::Op(Operator::Equal))?;

    let value = match lexer.peek().transpose()? {
        Some(token) => match token.kind {
            Kind::Op(Operator::LeftBrace) => parse_expr_block(lexer)?,
            _ => parse_expression(lexer)?,
        },
        _ => unreachable!(),
    };

    lexer.expect(Kind::Op(Operator::SemiColon))?;

    let location = Location::new(keyword.location.start_byte, value.location().end_byte);
    Ok(Expression::Var {
        mutable,
        name,
        value: Box::new(value),
        location,
    })
}

//fn parse_if_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, Error> {}

fn parse_with_precedence<'parser>(
    lexer: &mut Lexer<'parser>,
    min_precedence: u8,
) -> Result<Expression<'parser>, Error> {
    let asd = lexer.peek().transpose()?;
    let mut left = match asd {
        Some(token) => match &token.kind {
            Kind::Value(ref value) => match value {
                Value::Primitive(_) => parse_primitive(lexer)?,
                Value::Ident(_) => {
                    let (expr, _) = parse_identifier(lexer)?;
                    expr
                }
                t => todo!("{t}"),
            },
            Kind::Op(ref op) => match op {
                Operator::LeftParen => {
                    lexer.next().transpose()?;
                    let left = parse_with_precedence(lexer, precedences::BASE)?;
                    lexer.expect(Kind::Op(Operator::RightParen))?;
                    left
                }
                t => todo!("{t:?}"),
            },
            Kind::Return => parse_return_expression(lexer)?,
            //Kind::If => parse_if_expression(lexer)?,
            t => todo!("{t:?}"),
        },
        None => todo!(),
    };

    loop {
        let Some(next) = lexer.peek().transpose()? else {
            return Ok(left);
        };

        let Kind::Op(operator) = next.kind else {
            return Ok(left);
        };

        let precedence = get_precedence(operator);

        if precedence <= min_precedence {
            break;
        }

        let Some(_) = lexer.next().transpose()? else {
            unreachable!();
        };

        let right = parse_with_precedence(lexer, precedence)?;

        let location = Location::new(left.location().start_byte, right.location().end_byte);
        left = Expression::BinaryOp {
            lhs: Box::new(left),
            operator,
            rhs: Box::new(right),
            location,
        };
    }

    Ok(left)
}

fn parse_return_expression<'parser>(
    lexer: &mut Lexer<'parser>,
) -> Result<Expression<'parser>, Error> {
    let keyword = lexer.expect(Kind::Return)?;

    let value = parse_expression(lexer)?;

    let ending_semi = lexer.expect(Kind::Op(Operator::SemiColon))?;

    let location = keyword.location.start_byte..ending_semi.location.end_byte;
    Ok(Expression::Return {
        value: Box::new(value),
        location: location.into(),
    })
}

fn parse_primitive<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, Error> {
    let (primitive, location) = match lexer.next().transpose()? {
        Some(Token {
            kind: Kind::Value(Value::Primitive(primitive)),
            location,
            ..
        }) => (primitive, location),
        _ => unreachable!(),
    };

    match primitive {
        Primitive::Int { value, size } => Ok(Expression::IntLiteral {
            value,
            size,
            location,
        }),
        Primitive::UInt { value, size } => Ok(Expression::UintLiteral {
            value,
            size,
            location,
        }),
        Primitive::Float { value, size } => Ok(Expression::FloatLiteral {
            value,
            size,
            location,
        }),
        Primitive::Bool(value) => Ok(Expression::Bool { value, location }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Parser;

    fn make_sut(source: &str) -> Parser<'_> {
        Parser::new(source)
    }

    #[test]
    fn simple_math() {
        let simple_math_expr = "1 + 2 * 3 - 4";

        let mut parser = make_sut(simple_math_expr);

        let math_expr_ast = match parse_expression(&mut parser.lexer) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(math_expr_ast);
    }

    #[test]
    fn mutable_variable() {
        let variables = "var hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn immutable_constant() {
        let variables = "const hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn if_statement() {
        let source = r#"
            if something == another_thing {
                let this_is_a_var = 10 + 3;
            } else if something == 10 {
                let this_is_another = 10 + 10;
            } else {
                let omg = 1 + 1;
            }
        "#;
        let mut parser = make_sut(source);

        //let if_ast = match parse_expression(&mut parser.lexer) {
        //    Ok(expr) => expr,
        //    Err(e) => panic!("{e:?}"),
        //};

        //insta::assert_debug_snapshot!(if_ast);
    }
}
