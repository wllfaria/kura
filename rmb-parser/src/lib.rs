mod expression;

use rmb_lexer::{
    token::{FloatSizes, IntSizes, Kind, Location, Operator, Token, UIntSizes},
    Lexer, TransposeRef,
};

use crate::expression::{parse_expression, parse_identifier};
use miette::{Error, LabeledSpan, SourceSpan};

#[derive(Debug)]
pub enum Statement<'ast> {
    FunArgument {
        name: &'ast str,
        arg_type: Box<Expression<'ast>>,
        location: Location,
    },
    Fun {
        name: &'ast str,
        arguments: Vec<Statement<'ast>>,
        body: Vec<Expression<'ast>>,
        return_type: Option<Box<Expression<'ast>>>,
        location: Location,
    },
}

#[derive(Debug)]
pub enum Expression<'ast> {
    Var {
        mutable: bool,
        name: &'ast str,
        value: Box<Expression<'ast>>,
        location: Location,
    },
    Bool {
        value: bool,
        location: Location,
    },
    If {
        condition: Box<Expression<'ast>>,
        location: Location,
        body: Vec<Expression<'ast>>,
    },
    Ident {
        name: &'ast str,
        location: Location,
    },
    Block {
        expressions: Vec<Expression<'ast>>,
        location: Location,
    },
    UintLiteral {
        value: u64,
        size: Option<UIntSizes>,
        location: Location,
    },
    FloatLiteral {
        value: f64,
        size: Option<FloatSizes>,
        location: Location,
    },
    IntLiteral {
        value: i64,
        size: Option<IntSizes>,
        location: Location,
    },
    BinaryOp {
        operator: Operator,
        lhs: Box<Expression<'ast>>,
        rhs: Box<Expression<'ast>>,
        location: Location,
    },
    Return {
        value: Box<Expression<'ast>>,
        location: Location,
    },
}

impl Expression<'_> {
    fn location(&self) -> Location {
        match self {
            Expression::Var { location, .. } => *location,
            Expression::If { location, .. } => *location,
            Expression::Ident { location, .. } => *location,
            Expression::Bool { location, .. } => *location,
            Expression::Block { location, .. } => *location,
            Expression::UintLiteral { location, .. } => *location,
            Expression::Return { location, .. } => *location,
            Expression::FloatLiteral { location, .. } => *location,
            Expression::IntLiteral { location, .. } => *location,
            Expression::BinaryOp { location, .. } => *location,
        }
    }
}

pub struct Parser<'par> {
    pub source: &'par str,
    pub lexer: Lexer<'par>,
}

impl<'par> Parser<'par> {
    pub fn new(input: &'par str) -> Self {
        Self {
            source: input,
            lexer: Lexer::new(input),
        }
    }

    pub fn parse(mut self) -> Result<Vec<Statement<'par>>, Error> {
        let mut statements = vec![];

        loop {
            if self.lexer.is_empty() {
                break;
            }

            let statement = self.parse_statement()?;
            statements.push(statement);
        }

        Ok(statements)
    }

    fn parse_statement(&mut self) -> Result<Statement<'par>, Error> {
        match self.lexer.peek().transpose()? {
            Some(token) => match &token.kind {
                Kind::Fun => self.parse_function(),
                t => todo!("{t:?}"),
            },
            _ => todo!(),
        }
    }

    fn parse_function_args(&mut self, start_kw: &Token<'_>) -> Result<Vec<Statement<'par>>, Error> {
        let mut arguments = vec![];

        loop {
            let (arg_name_expr, arg_name) = parse_identifier(&mut self.lexer)?;
            self.lexer.expect(Kind::Op(Operator::Colon))?;
            let (arg_type, _) = parse_identifier(&mut self.lexer)?;

            arguments.push(Statement::FunArgument {
                name: arg_name,
                location: Location::new(
                    arg_name_expr.location().start_byte,
                    arg_type.location().end_byte,
                ),
                arg_type: Box::new(arg_type),
            });

            match self.lexer.peek().transpose()? {
                Some(token) => match token.kind {
                    Kind::Op(Operator::RightParen) => break,
                    Kind::Op(Operator::Comma) => (),
                    _ => {
                        let location = start_kw.location.start_byte..token.location.end_byte;
                        return Err(miette::miette! {
                            labels = vec![
                                LabeledSpan::at(SourceSpan::from(location), "this function declaration")
                            ],
                            help = "did you forget to add a comma (,) here?",
                            "syntax error: unexpected identifier after argument type"
                        }
                        .with_source_code(self.source.to_string()));
                    }
                },
                None => break,
            }
        }

        Ok(arguments)
    }

    fn parse_function(&mut self) -> Result<Statement<'par>, Error> {
        let keyword = self.lexer.expect(Kind::Fun)?;
        let (_, fun_name) = parse_identifier(&mut self.lexer)?;
        self.lexer.expect(Kind::Op(Operator::LeftParen))?;

        let arguments = self.parse_function_args(&keyword)?;

        // after parsing argument list we need to consume the closing parenthesis
        self.lexer.expect(Kind::Op(Operator::RightParen))?;

        // after the argument list of a function, there can be an optional return type annotation
        // => <TYPE> {
        // before the left brace, but its fine to be ommited
        let has_return = match self.lexer.peek().transpose()? {
            Some(token) => matches!(token.kind, Kind::Op(Operator::ThickArrow)),
            _ => false,
        };

        let return_type = if has_return {
            self.lexer.next().transpose()?;
            let (identifier, _) = parse_identifier(&mut self.lexer)?;
            Some(Box::new(identifier))
        } else {
            None
        };

        self.lexer.expect(Kind::Op(Operator::LeftBrace))?;

        let mut body = vec![];

        loop {
            match self.lexer.peek().transpose()? {
                Some(token) => {
                    if let Kind::Op(Operator::RightBrace) = token.kind {
                        break;
                    }
                }
                None => break,
            }
            body.push(parse_expression(&mut self.lexer)?);
        }

        // consume the closing brace of the function
        let closing_brace = self.lexer.expect(Kind::Op(Operator::RightBrace))?;

        let location = keyword.location.start_byte..closing_brace.location.end_byte;
        Ok(Statement::Fun {
            name: fun_name,
            arguments,
            body,
            return_type,
            location: location.into(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Parser<'_> {
        Parser::new(source)
    }

    #[test]
    fn function_declaration() {
        let source = r#"
            fun calculate_circumference(diameter: f64) => f64 {
                const pi = 3.14159265358979323846264338327950288_f32;
                const radius = diameter / 2.0;
                const circumference = 2.0 * pi * radius;

                const nesting = {
                    const something = 10;
                    var nesting_more = {
                        return 10 + 3 * 4;
                    };
                    return 10 + something;
                };
            
                return circumference;
            }"#;

        let ast = match make_sut(source).parse() {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(ast);
    }
}
