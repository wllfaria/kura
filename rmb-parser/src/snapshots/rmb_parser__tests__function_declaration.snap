---
source: rmb-parser/src/lib.rs
expression: ast
---
[
    Fun {
        name: "calculate_circumference",
        arguments: [
            FunArgument {
                name: "diameter",
                arg_type: Ident {
                    name: "f64",
                    location: Location {
                        start_byte: 51,
                        end_byte: 54,
                    },
                },
                location: Location {
                    start_byte: 41,
                    end_byte: 54,
                },
            },
        ],
        body: [
            Var {
                mutable: false,
                name: "pi",
                typ: None,
                value: FloatLiteral {
                    value: 3.141592653589793,
                    size: Some(
                        F32,
                    ),
                    location: Location {
                        start_byte: 92,
                        end_byte: 133,
                    },
                },
                location: Location {
                    start_byte: 81,
                    end_byte: 133,
                },
            },
            Var {
                mutable: false,
                name: "radius",
                typ: None,
                value: BinaryOp {
                    operator: Slash,
                    lhs: Ident {
                        name: "diameter",
                        location: Location {
                            start_byte: 166,
                            end_byte: 174,
                        },
                    },
                    rhs: FloatLiteral {
                        value: 2.0,
                        size: None,
                        location: Location {
                            start_byte: 177,
                            end_byte: 180,
                        },
                    },
                    location: Location {
                        start_byte: 166,
                        end_byte: 180,
                    },
                },
                location: Location {
                    start_byte: 151,
                    end_byte: 180,
                },
            },
            Var {
                mutable: false,
                name: "circumference",
                typ: None,
                value: BinaryOp {
                    operator: Star,
                    lhs: BinaryOp {
                        operator: Star,
                        lhs: FloatLiteral {
                            value: 2.0,
                            size: None,
                            location: Location {
                                start_byte: 220,
                                end_byte: 223,
                            },
                        },
                        rhs: Ident {
                            name: "pi",
                            location: Location {
                                start_byte: 226,
                                end_byte: 228,
                            },
                        },
                        location: Location {
                            start_byte: 220,
                            end_byte: 228,
                        },
                    },
                    rhs: Ident {
                        name: "radius",
                        location: Location {
                            start_byte: 231,
                            end_byte: 237,
                        },
                    },
                    location: Location {
                        start_byte: 220,
                        end_byte: 237,
                    },
                },
                location: Location {
                    start_byte: 198,
                    end_byte: 237,
                },
            },
            Var {
                mutable: false,
                name: "nesting",
                typ: None,
                value: Block {
                    expressions: [
                        Var {
                            mutable: false,
                            name: "something",
                            typ: None,
                            value: UintLiteral {
                                value: 10,
                                size: None,
                                location: Location {
                                    start_byte: 312,
                                    end_byte: 314,
                                },
                            },
                            location: Location {
                                start_byte: 294,
                                end_byte: 314,
                            },
                        },
                        Var {
                            mutable: true,
                            name: "nesting_more",
                            typ: None,
                            value: Block {
                                expressions: [
                                    Return {
                                        value: BinaryOp {
                                            operator: Plus,
                                            lhs: UintLiteral {
                                                value: 10,
                                                size: None,
                                                location: Location {
                                                    start_byte: 388,
                                                    end_byte: 390,
                                                },
                                            },
                                            rhs: BinaryOp {
                                                operator: Star,
                                                lhs: UintLiteral {
                                                    value: 3,
                                                    size: None,
                                                    location: Location {
                                                        start_byte: 393,
                                                        end_byte: 394,
                                                    },
                                                },
                                                rhs: UintLiteral {
                                                    value: 4,
                                                    size: None,
                                                    location: Location {
                                                        start_byte: 397,
                                                        end_byte: 398,
                                                    },
                                                },
                                                location: Location {
                                                    start_byte: 393,
                                                    end_byte: 398,
                                                },
                                            },
                                            location: Location {
                                                start_byte: 388,
                                                end_byte: 398,
                                            },
                                        },
                                        location: Location {
                                            start_byte: 381,
                                            end_byte: 399,
                                        },
                                    },
                                ],
                                location: Location {
                                    start_byte: 355,
                                    end_byte: 421,
                                },
                            },
                            location: Location {
                                start_byte: 336,
                                end_byte: 421,
                            },
                        },
                        Return {
                            value: BinaryOp {
                                operator: Plus,
                                lhs: UintLiteral {
                                    value: 10,
                                    size: None,
                                    location: Location {
                                        start_byte: 450,
                                        end_byte: 452,
                                    },
                                },
                                rhs: Ident {
                                    name: "something",
                                    location: Location {
                                        start_byte: 455,
                                        end_byte: 464,
                                    },
                                },
                                location: Location {
                                    start_byte: 450,
                                    end_byte: 464,
                                },
                            },
                            location: Location {
                                start_byte: 443,
                                end_byte: 465,
                            },
                        },
                    ],
                    location: Location {
                        start_byte: 272,
                        end_byte: 483,
                    },
                },
                location: Location {
                    start_byte: 256,
                    end_byte: 483,
                },
            },
            Return {
                value: Ident {
                    name: "circumference",
                    location: Location {
                        start_byte: 521,
                        end_byte: 534,
                    },
                },
                location: Location {
                    start_byte: 514,
                    end_byte: 535,
                },
            },
        ],
        return_type: Some(
            Ident {
                name: "f64",
                location: Location {
                    start_byte: 59,
                    end_byte: 62,
                },
            },
        ),
        location: Location {
            start_byte: 13,
            end_byte: 549,
        },
    },
]
